base/cvd/cuttlefish/host/commands/cvd/selector/instance_record.cpp

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (C) 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "host/commands/cvd/selector/instance_record.h"
+
+#include <android-base/logging.h>
+#include <fmt/format.h>
+
+namespace cuttlefish {
+namespace selector {
+
+namespace {
+constexpr int BASE_ADB_PORT = 6520;
+constexpr int BASE_INSTANCE_ID = 1;
+}  // namespace
+
+LocalInstance::LocalInstance(std::shared_ptr<cvd::InstanceGroup> group_proto,
+                             cvd::Instance* instance_proto)
+    : group_proto_(group_proto), instance_proto_(instance_proto) {}
+
+void LocalInstance::set_state(cvd::InstanceState state) {
+  instance_proto_->set_state(state);
+}
+
+std::string LocalInstance::instance_dir() const {
+  return fmt::format("{}/cuttlefish/instances/cvd-{}",
+                     group_proto_->home_directory(), id());
+}
+
+int LocalInstance::adb_port() const {
+  // run_cvd picks this port from the instance id and doesn't provide a flag
+  // to change in cvd_internal_flag
+  return id() > 0 ? id() + BASE_ADB_PORT - BASE_INSTANCE_ID : 0;
```

Comment:
nit: I think handling the zero  scenario as an early return makes it an easier reading in this case.

```
// TODO: This should never happen.
if id() == 0 {
  return 0;
}
```

also, the following order makes easier to read as well:  ` BASE_ADB_PORT + (id() - BASE_INSTANCE_ID)`